Add recvmsg() and pktinfo support

jukkar opened the pull request:
This PR adds support for `recvmsg()` socket call. It also adds support for `IP_PKTINFO` and `IPV6_RECVPKTINFO` socket options. Tests for both of these features are also added.

Fixes #36415

jukkar committed:
net: context: Add support for setting receive pktinfo option

Add low level support for setting IP_PKTINFO or IPV6_RECVPKTINFO
socket options. The support is disabled by default.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Add support for setting pktinfo options

Add IP_PKTINFO or IPV6_RECVPKTINFO BSD socket options that
can be used to get extra information of received data in
the ancillary data in recvmsg() call.

For IPV6_RECVPKTINFO see RFC 3542 for details.
For IP_PKTINFO see Linux ip(7) manual page for details.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Make receive timeout shorter

No need to wait 2.5 seconds for the test to pass, 400ms wait
is enough.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: sockets: Remove extra check from sendmsg()

There was double "if (status < 0)" check in sendmsg(),
remove the extra check.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: sockets: Add recvmsg() implementation

Add support for recvmsg() function which can return data
in msghdr struct (iovec).

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: Add UDP tests for recvmsg()

Make sure recvmsg() works as expected for UDP data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: Add TCP tests for recvmsg()

Make sure recvmsg() works as expected for TCP data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Add support for filling receive pktinfo data

If user has set either IP_PKTINFO (for IPv4) or
IPV6_RECVPKTINFO (for IPv6) socket options, then the system
will return relevant information in recvmsg() ancillary data.

Fixes #36415

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: Add recv pktinfo ancillary tests for recvmsg

Make sure recvmsg() is able to return IPv4 IP_PKTINFO and
IPv6 IPV6_RECVPKTINFO ancillary data to the caller.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1539,6 +1634,205 @@ ssize_t z_vrfy_zsock_recvfrom(int sock, void *buf, size_t max_len, int flags,
+		goto fail;
+	}
+
+	for (i = 0; i < iovlen; i++) {
```
Shall we have an initial loop run over allocated `msg_iov` entries, to clear `iov_base` pointer? Right now, in case `msg_iovlen > 1` and let's say allocation for the first `msg_iov` entry fails, we still have remaining `iov_base` pointer copies from the original msghdr. This will cause trouble when we jump to fail handler (will try to release the memory we did not allocate here).

jukkar replied:
Yes, good point.

rlubos commented on subsys/net/ip/net_context.c:
```diff
@@ -1558,6 +1569,32 @@ static int get_context_ipv6_v6only(struct net_context *context,
+	}
+
+	if (context->options.recv_pktinfo == true) {
+		*((int *)value) = (int) true;
```
nit: No need to add space after cast.

Why not cast `context->options.recv_pktinfo` directly, w/o if statement?

jukkar replied:
There are now multiple get option functions in this file, they all follow this same pattern. I think we could create a generic function for these, but in separate PR. I can prepare one after this one is merged.

rlubos commented on include/zephyr/net/socket.h:
```diff
@@ -1095,10 +1095,33 @@ struct ifreq {
+#define IPV6_RECVPKTINFO 49
+
+struct in6_pktinfo {
+	struct in6_addr ipi6_addr;    /* src/dst IPv6 address */
```
When is it src and when is it dst address? It seems we only set this as a dst address on receive?

jukkar replied:
I copied the struct verbatim from the RFC. I think the struct can be used in other ways too (did not check this).

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1191,6 +1191,88 @@ int zsock_wait_data(struct net_context *ctx, k_timeout_t *timeout)
+	cmsg->cmsg_level = level;
+	cmsg->cmsg_type = type;
+
+	memcpy(CMSG_DATA(cmsg), pktinfo, pktinfo_len);
```
Shall `msg->msg_controllen` be verified to be large enough before copy?

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1191,6 +1191,88 @@ int zsock_wait_data(struct net_context *ctx, k_timeout_t *timeout)
+{
+	struct cmsghdr *cmsg;
+
+	cmsg = CMSG_FIRSTHDR(msg);
```
Shall `cmsg` be checked for NULL ptr? `CMSG_FIRSTHDR` will return NULL if `msg->msg_controllen < sizeof(struct cmsghdr)`

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1416,11 +1477,44 @@ static ssize_t zsock_recv_stream_timed(struct net_context *ctx, uint8_t *buf, si
+				available_len = msg->msg_iov[iovec].iov_len;
+				msg->msg_iov[iovec].iov_len = 0;
+
+				continue;
```
Should it also be checked if there is any remaining data on the rx queue before `continue`? Because right now, if the data available on the stream matched exactly `iov_len`, and there is more `msg_iov` available, we're getting back to `zsock_wait_data()`, potentially blocking while the function could return what was received so far.

rlubos reviewed (commented):
Nice, some parts seem really tricky so I'm even more impressed you've managed to implement this in such a short time.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -1539,6 +1634,205 @@ ssize_t z_vrfy_zsock_recvfrom(int sock, void *buf, size_t max_len, int flags,
+		}
+	}
+
+	k_usermode_from_copy(&msg_copy.msg_flags,
```
Yes it is copied, will fix.

pdgendt reviewed (commented):
There is one compliance issue to be fixed and some nits from me.

jukkar committed:
net: socket: Return ENOTSUP for unknown socket type in recvfrom()

If we for some reason are supplied unsupported socket type in
recvfrom(), then return ENOTSUP error to the caller instead of
silently accept it by returning 0.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos reviewed (dismissed):
LGTM, thanks

jukkar commented:
Fixing the pktinfo tests so that we always catch if the control buffer is too small.
